python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+The core of this is a Matlab "gateway" function, [`call_method.m`](https://github.com/mducle/hugo/blob/master/src/call_method.m#L101)
+which takes the name of a Matlab function and its arguments and uses `feval` to evaluate it.
+When compiled using the Matlab Compiler SDK as a Python module, it maybe accessed in Python using `<module_name>.call_method('function_name', object, [<arg_list>])`.
+To make it easier to use, this core gateway function and the Matlab Python module (called `horace`) is wrapped in a Python class,
+[`pyHorace.Matlab`](https://github.com/mducle/hugo/blob/master/pyHorace/Matlab.py#L60) whose `__getattr__()` special method is overloaded to call `call_method`.
+That is, when the user does:
+
+```
+from pyHorace import Matlab
+Matlab.<func_name>(*args)
+```
+
+The `Matlab.<func_name>` part causes Python to call `Matlab.__getattr__('func_name')`. 
+This returns a Python function which wraps `horace.call_method('func_name', [], *args)`.
+Both these steps are chained by Python so that uses can do `Matlab.cut_sqw('sqw_file', [], [], [], [-1,1])` transparently.
+Because the wrapper to `call_method` is through the `__getattr__` overload of a class, users always have to use the "dot" method-access notation.
+I.e., `from pyHorace.Matlab import *; cut_sqw('sqw_file', [], [], [], [-1,1])` will not work because `pyHorace.Matlab` is not a Python module but a class.
+In addition `help(pyHorace.Matlab.cut_sqw)` will also not work because this actually refers to a (dynamically constructed) Python wrapper function.
+Thus if we want the above two functionalities (accessing Matlab "included" functions/classes and help)
+we will need to write explicit Python wrappers for *every* Matlab function we want to expose this way.
+
+The other major feature of `pyHorace` is a Python [`MatlabProxyObject`](https://github.com/mducle/hugo/blob/master/pyHorace/MatlabProxyObject.py) class
+which wraps a Matlab class so that it appears transparent to Python users.
+That is, Python users can call the methods and access the properties of any Matlab class in Python using the "dot" notation as if they are in Matlab.
+This works by overloading the special Python methods `__getattr__` and `__setattr__`
+to call the equivalent Matlab special functions `subsref` and `subassgn`
+to access the methods/properties of a class.
+In addition, the Python `__repr__`, `__str__`, `__dir__` and `__doc__` methods are also overloaded
+to provide a description of the object, its methods and help text (except that the Matlab help system is not accessible in a compiled application).
+
+This wrapper is needed because any new-style (defined by `classdef`) Matlab object returned by `call_method` and its wrappers
+is opaque to Python and cannot be handled by purely Python functions.
+Old-style Matlab objects are treated as plain `struct`s and are automatically converted to Python `dict`s,
```

Comment:
The ongoing Horace refactor should be replacing all of the old-style classes with equivalent new-style ones; however having support for old-style classes is probably not a bad addition.